Modules/_zstd/_zstdmodule.c

```diff
@@ -0,0 +1,968 @@
+                          int compression_level)
+/*[clinic end generated code: output=9c2a7d8c845cee93 input=ef6f4f2b0b7e178d]*/
+{
+    if (ZSTD_versionNumber() < 10405) {
+        /* Must be dynamically linked */
+        PyErr_Format(PyExc_NotImplementedError,
+                "The _finalize_dict function is only available when the underlying "
+                "zstd library's version is greater than or equal to v1.4.5. "
+                "The current zstd version is v%s.",
+                ZSTD_versionString());
+        return NULL;
+    }
+
+    Py_ssize_t chunks_number;
+    size_t *chunk_sizes = NULL;
+    PyObject *dst_dict_bytes = NULL;
+    size_t zstd_ret;
+    ZDICT_params_t params;
+    Py_ssize_t sizes_sum;
+    Py_ssize_t i;
+
+    /* Check arguments */
+    if (dict_size <= 0) {
+        PyErr_SetString(PyExc_ValueError, "dict_size argument should be positive number.");
+        return NULL;
+    }
+
+    chunks_number = Py_SIZE(samples_size_list);
+    if ((size_t) chunks_number > UINT32_MAX) {
+        PyErr_SetString(PyExc_ValueError,
+                        "The number of samples should <= UINT32_MAX.");
+        return NULL;
+    }
+
+    /* Prepare chunk_sizes */
+    chunk_sizes = PyMem_Malloc(chunks_number * sizeof(size_t));
+    if (chunk_sizes == NULL) {
+        PyErr_NoMemory();
+        goto error;
+    }
+
+    sizes_sum = 0;
+    for (i = 0; i < chunks_number; i++) {
+        PyObject *size = PyList_GET_ITEM(samples_size_list, i);
+        chunk_sizes[i] = PyLong_AsSize_t(size);
+        if (chunk_sizes[i] == (size_t)-1 && PyErr_Occurred()) {
+            PyErr_SetString(PyExc_ValueError,
+                            "Items in samples_size_list should be an int "
+                            "object, with a size_t value.");
+            goto error;
+        }
+        sizes_sum += chunk_sizes[i];
+    }
+
+    if (sizes_sum != Py_SIZE(samples_bytes)) {
+        PyErr_SetString(PyExc_ValueError,
+                        "The samples size list doesn't match the concatenation's size.");
+        goto error;
+    }
+
+    /* Allocate dict buffer */
+    dst_dict_bytes = PyBytes_FromStringAndSize(NULL, dict_size);
+    if (dst_dict_bytes == NULL) {
+        goto error;
+    }
+
+    /* Parameters */
+
+    /* Optimize for a specific zstd compression level, 0 means default. */
+    params.compressionLevel = compression_level;
+    /* Write log to stderr, 0 = none. */
+    params.notificationLevel = 0;
+    /* Force dictID value, 0 means auto mode (32-bits random value). */
+    params.dictID = 0;
+
+    /* Finalize the dictionary */
+    Py_BEGIN_ALLOW_THREADS
+    zstd_ret = ZDICT_finalizeDictionary(
+                        PyBytes_AS_STRING(dst_dict_bytes), dict_size,
+                        PyBytes_AS_STRING(custom_dict_bytes), Py_SIZE(custom_dict_bytes),
+                        PyBytes_AS_STRING(samples_bytes), chunk_sizes,
+                        (uint32_t)chunks_number, params);
+    Py_END_ALLOW_THREADS
+
```

Comment:
And from here until the end, the logic is also the same. Can we also refactor this into a separate function?